File: CoinChange_minCoins.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        arr = [[None for _ in range(0,amount+1)] for _ in range(len(coins)+1)]
        for i in range(len(arr)):
            arr[i][0]=0
        for i in range(1,len(arr[0])):
            arr[0][i]=amount+1
        for i in range(1, len(arr)):
            for j in range(1, len(arr[0])):
                if coins[i-1]>j:
                    arr[i][j]=arr[i-1][j]
                else:
                    arr[i][j]=min(arr[i-1][j], 1+arr[i][j-coins[i-1]])
        result = arr[len(coins)][amount]
        if result>amount:
            return -1
        return result
        # print(arr)
        # [print(i) for i in len(arr)]
        
#         if not coins:
#             return 0
#         return self.helper(coins, amount, 0, 0)
    
#     def helper(self, coins:List[int], amount:int, index:int, mincoins:int):
#         if amount==0:
#             return mincoins
#         elif amount<0 or index>=len(coins):
#             return -1
        
#         case1 = self.helper(coins, amount-coins[index], index, mincoins+1)
#         case2 = self.helper(coins, amount, index+1, mincoins)
#         if case1==-1:
#             return case2
#         if case2==-1:
#             return case1
#         return min(case1, case2)


""" Time complexity - O(nm) where n is the number of coins and m is the amount given
# space complexity - O(nm) as we are utilizing array as the intermediate data structure for dynamic programming.

For the commented recursion code, the time complexity would be 2^n 
and space complexity O(n) where n is the number of coins"""
        

File: CoinChangeMinCoins.java
// Time Complexity : O(nm) //iterating over entire nums array : Optionation for previous exponential solutions
// Space Complexity : O(mxn) // m = amount, n = total different denomination coins
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : Learning DP. Able to code post watching videos. Please validate time complexity.

/*
Built two dimension array for Repeated sub problem. Based on that computation came up with this solution.
We have 2 route choose coin or not. Calculate minimum at every step.
*/

public class CoinChangeMinCoins {
    public int coinChange(int[] coins, int amount) {
        if(coins == null || coins.length == 0)
            return -1;

        int n = coins.length;
        int m = amount;

        int[][] dp = new int[n+1][m+1];
        dp[0][0] = 0;
        for(int j=1; j<=m; j++){
//          dp[1][j] = Integer.MAX_VALUE;
            dp[0][j] = 99999;
        }

        for(int i=1; i<=n ; i++) {
            for(int j=1; j<=m; j++){
                if(j < coins[i-1]) {
                    dp[i][j] = dp[i-1][j];
                }else{
                    dp[i][j] = Math.min(dp[i-1][j] , 1+ dp[i][j-coins[i-1]]);
                }

            }
        }
        int result = dp[n][m];
        if(result>= 99999) return -1;

        return result;
    }
}